File: duplex/go/wire.go
package duplex

import (
	"errors"
	"sort"
	"strconv"
	"strings"
	"sync"
	"unicode/utf8"

	bitwire "github.com/Bitspark/bitwire/wire/go"
)

var (
	ErrPath           = errors.New("invalid wire path")
	ErrNoRoute        = errors.New("wire path has no destination")
	ErrReceiverExists = errors.New("endpoint already has a receiver")
)

// EncodePath concatenates UTF-8 byte-length-prefixed scalar-string segments.
// The empty path is "", while a single empty segment is "0:".
func EncodePath(path []string) (string, error) {
	var encoded strings.Builder
	for _, segment := range path {
		if !utf8.ValidString(segment) {
			return "", ErrPath
		}
		encoded.WriteString(strconv.Itoa(len(segment)))
		encoded.WriteByte(':')
		encoded.WriteString(segment)
	}
	return encoded.String(), nil
}

// DecodePath accepts only the canonical form of EncodePath, without Unicode
// normalization or interpretation of dots, slashes or empty segments.
func DecodePath(encoded string) ([]string, error) {
	path := []string{}
	for encoded != "" {
		colon := strings.IndexByte(encoded, ':')
		if colon <= 0 {
			return nil, ErrPath
		}
		digits := encoded[:colon]
		if len(digits) > 1 && digits[0] == '0' {
			return nil, ErrPath
		}
		for _, digit := range digits {
			if digit < '0' || digit > '9' {
				return nil, ErrPath
			}
		}
		length, err := strconv.ParseUint(digits, 10, 64)
		encoded = encoded[colon+1:]
		if err != nil || length > uint64(len(encoded)) {
			return nil, ErrPath
		}
		segment := encoded[:int(length)]
		if !utf8.ValidString(segment) {
			return nil, ErrPath
		}
		path = append(path, segment)
		encoded = encoded[int(length):]
	}
	return path, nil
}

type selectedWire struct {
	root   bitwire.Wire
	prefix []string
}

// At selects a relative path without allocating a peer, channel or queue.
// The selection grants only send access, even when path is empty.
func At(root bitwire.Wire, path []string) bitwire.Wire {
	return &selectedWire{root: root, prefix: append([]string{}, path...)}
}

func (w *selectedWire) path(path []string) []string {
	return append(append([]string{}, w.prefix...), path...)
}
func (w *selectedWire) Send(path []string, message bitwire.Message) error {
	return w.root.Send(w.path(path), message)
}

type mountedWire struct {
	children map[string]bitwire.Endpoint
	mu       sync.Mutex
	closed   bool
	current  *mountedReceiver
}
type mountedReceiver struct {
	receiver  bitwire.Receiver
	active    bool
	children  []*mountedChild
	remaining int
}
type mountedChild struct {
	detach func()
	ended  bool
}

// Mount consumes one path segment and delegates to that child. The map is
// copied. A mount has no leaf at []; [""] can select an empty-string key.
// Its single receive attachment borrows one attachment from each child.
// Closing a mount detaches those attachments and leaves every child usable.
func Mount(children map[string]bitwire.Endpoint) bitwire.Endpoint {
	w := &mountedWire{children: make(map[string]bitwire.Endpoint, len(children))}
	for key, child := range children {
		w.children[key] = child
	}
	return w
}

func (w *mountedWire) destination(path []string) (bitwire.Endpoint, error) {
	if w.closed {
		return nil, ErrClosed
	}
	if len(path) == 0 {
		return nil, ErrNoRoute
	}
	if _, err := EncodePath(path); err != nil {
		return nil, err
	}
	child := w.children[path[0]]
	if child == nil {
		return nil, ErrNoRoute
	}
	return child, nil
}
func (w *mountedWire) Send(path []string, message bitwire.Message) error {
	w.mu.Lock()
	child, err := w.destination(path)
	w.mu.Unlock()
	if err != nil {
		return err
	}
	return child.Send(append([]string{}, path[1:]...), message)
}
func (w *mountedWire) Receive(receiver bitwire.Receiver) (func(), error) {
	w.mu.Lock()
	if w.closed {
		w.mu.Unlock()
		return nil, ErrClosed
	}
	if w.current != nil {
		w.mu.Unlock()
		return nil, ErrReceiverExists
	}
	keys := make([]string, 0, len(w.children))
	for key, child := range w.children {
		if child != nil {
			keys = append(keys, key)
		}
	}
	sort.Strings(keys)
	attachment := &mountedReceiver{receiver: receiver, active: true, remaining: len(keys)}
	for range keys {
		attachment.children = append(attachment.children, &mountedChild{})
	}
	w.current = attachment
	w.mu.Unlock()

	for i, key := range keys {
		slot := attachment.children[i]
		w.mu.Lock()
		active := attachment.active
		w.mu.Unlock()
		if !active {
			return nil, ErrClosed
		}
		detach, err := w.children[key].Receive(bitwire.Receiver{
			Message: func(path []string, message bitwire.Message) {
				// The child owns capture of accepted invocations. A retained
				// delivery, including cancellation, keeps its original receiver.
				if receiver.Message != nil {
					receiver.Message(append([]string{key}, path...), message)
				}
			},
			Closed: func(code bitwire.Code, reason string) { w.childEnded(attachment, slot, code, reason) },
		})
		w.mu.Lock()
		active = attachment.active && !slot.ended
		if err == nil && active {
			slot.detach = detach
		}
		w.mu.Unlock()
		if err != nil || !active {
			// Close may happen while the child's Receive is returning. Its
			// late disposer is still ours, even after the attachment ended.
			if detach != nil {
				detach()
			}
			w.remove(attachment)
			if err != nil {
				return nil, err
			}
			return nil, ErrClosed
		}
	}
	w.mu.Lock()
	active := attachment.active
	w.mu.Unlock()
	if !active {
		return nil, ErrClosed
	}
	return func() { w.remove(attachment) }, nil
}

// releaseLocked retires only this attachment. Clear its ownership before
// invoking borrowed disposers or callbacks, which may reenter the mount.
func (w *mountedWire) releaseLocked(attachment *mountedReceiver) []func() {
	attachment.active = false
	if w.current == attachment {
		w.current = nil
	}
	var detaches []func()
	for _, child := range attachment.children {
		if child.detach != nil {
			detaches = append(detaches, child.detach)
			child.detach = nil
		}
	}
	return detaches
}

func (w *mountedWire) remove(attachment *mountedReceiver) {
	w.mu.Lock()
	if !attachment.active {
		w.mu.Unlock()
		return
	}
	detaches := w.releaseLocked(attachment)
	w.mu.Unlock()
	for _, detach := range detaches {
		detach()
	}
}

func (w *mountedWire) childEnded(attachment *mountedReceiver, child *mountedChild, code bitwire.Code, reason string) {
	w.mu.Lock()
	if !attachment.active || child.ended {
		w.mu.Unlock()
		return
	}
	child.ended = true
	attachment.remaining--
	last := attachment.remaining == 0
	var detaches []func()
	if last {
		detaches = w.releaseLocked(attachment)
	} else if child.detach != nil {
		detaches = append(detaches, child.detach)
		child.detach = nil
	}
	w.mu.Unlock()
	for _, detach := range detaches {
		detach()
	}
	if last && attachment.receiver.Closed != nil {
		attachment.receiver.Closed(code, reason)
	}
}

func (w *mountedWire) Close(code bitwire.Code, reason string) error {
	w.mu.Lock()
	if w.closed {
		w.mu.Unlock()
		return nil
	}
	w.closed = true
	attachment := w.current
	var detaches []func()
	if attachment != nil {
		detaches = w.releaseLocked(attachment)
	}
	w.mu.Unlock()
	for _, detach := range detaches {
		detach()
	}
	if attachment != nil && attachment.receiver.Closed != nil {
		attachment.receiver.Closed(code, reason)
	}
	return nil
}

File: duplex/ts/src/wire.ts
import type { Endpoint, Path, Receiver, Wire } from '@bitspark/bitwire';

export class WireError extends Error {
  readonly code: 'closed' | 'no_route' | 'receiver_exists' | 'invalid_path';
  constructor(code: WireError['code']) {
    super(`Wire ${code}.`);
    this.name = 'WireError';
    this.code = code;
  }
}

function scalar(value: string): void {
  for (let i = 0; i < value.length; i++) {
    const unit = value.charCodeAt(i);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const low = value.charCodeAt(++i);
      if (!(low >= 0xdc00 && low <= 0xdfff)) throw new WireError('invalid_path');
    } else if (unit >= 0xdc00 && unit <= 0xdfff) throw new WireError('invalid_path');
  }
}

/** UTF-8 byte-length-prefixed segments; [] is '', whereas [''] is '0:'. */
export function encodePath(path: Path): string {
  const encoder = new TextEncoder();
  return path
    .map((segment) => {
      scalar(segment);
      return `${encoder.encode(segment).length}:${segment}`;
    })
    .join('');
}
/** Accepts only the canonical encoding, retaining dots, empty strings and BOMs. */
export function decodePath(encoded: string): string[] {
  scalar(encoded);
  const bytes = new TextEncoder().encode(encoded);
  const decoder = new TextDecoder('utf-8', { fatal: true, ignoreBOM: true });
  const path: string[] = [];
  for (let offset = 0; offset < bytes.length;) {
    const start = offset;
    let length = 0;
    while (offset < bytes.length && bytes[offset] !== 58) {
      const digit = bytes[offset++]! - 48;
      if (digit < 0 || digit > 9) throw new WireError('invalid_path');
      length = length * 10 + digit;
      if (!Number.isSafeInteger(length)) throw new WireError('invalid_path');
    }
    if (offset === start || offset === bytes.length || (offset - start > 1 && bytes[start] === 48))
      throw new WireError('invalid_path');
    offset++;
    if (length > bytes.length - offset) throw new WireError('invalid_path');
    try {
      path.push(decoder.decode(bytes.subarray(offset, offset + length)));
    } catch {
      throw new WireError('invalid_path');
    }
    offset += length;
  }
  return path;
}

/** Selects send access without granting receive attachment or closure authority. */
export function at(root: Wire, path: Path): Wire {
  const prefix = [...path];
  return {
    send: (suffix, message) => root.send([...prefix, ...suffix], message),
  };
}

interface ChildAttachment {
  active: boolean;
  detach?: () => void;
}
interface Attachment {
  receiver: Receiver;
  active: boolean;
  children: ChildAttachment[];
}
/**
 * Consumes one path segment; [] has no leaf, and [''] can select an empty key.
 * One owning receiver spans the borrowed children and sees their keys restored.
 * Copies the map. Closing detaches this mount's attachment, never children.
 */
export function mount(children: ReadonlyMap<string, Endpoint>): Endpoint {
  const routes = new Map(children);
  let attachment: Attachment | undefined;
  let closed = false;
  const destination = (path: Path): Endpoint => {
    if (closed) throw new WireError('closed');
    encodePath(path);
    const child = path.length ? routes.get(path[0]!) : undefined;
    if (!child) throw new WireError('no_route');
    return child;
  };
  const release = (child: ChildAttachment): void => {
    child.active = false;
    const detach = child.detach;
    child.detach = undefined;
    detach?.();
  };
  const remove = (held: Attachment, ending?: { code: number; reason: string }): void => {
    if (!held.active) return;
    held.active = false;
    if (attachment === held) attachment = undefined;
    for (const child of held.children) release(child);
    if (ending) held.receiver.closed?.(ending.code, ending.reason);
  };
  const receive = (receiver: Receiver): (() => void) => {
    if (closed) throw new WireError('closed');
    if (attachment) throw new WireError('receiver_exists');
    const held: Attachment = { receiver, active: true, children: [] };
    attachment = held;
    let remaining = routes.size;
    try {
      for (const [key, child] of routes) {
        const slot: ChildAttachment = { active: true };
        held.children.push(slot);
        const detach = child.receive({
          // The child's runtime owns admission and captured cancellation. Keep
          // its captured receiver even after this attachment is detached.
          message: (suffix, message) => receiver.message?.([key, ...suffix], message),
          closed: (code, reason) => {
            if (!held.active || !slot.active) return;
            remaining--;
            release(slot);
            if (remaining === 0) remove(held, { code, reason });
          },
        });
        // A child may synchronously end or close this mount while receiving.
        // Its returned disposer still belongs to this acquisition attempt.
        if (!held.active || !slot.active) {
          detach();
          throw new WireError('closed');
        }
        slot.detach = detach;
      }
    } catch (error) {
      remove(held);
      throw error;
    }
    return () => remove(held);
  };
  return {
    send: (path, message) => destination(path).send(path.slice(1), message),
    receive,
    close: (code = 1000, reason = '') => {
      if (closed) return;
      closed = true;
      if (attachment) remove(attachment, { code, reason });
    },
  };
}
